### PR TITLE
New Metric: Upper Face Dynamics Deviation(FDD)

### LINKED ISSUE #3097: Upper Face Dynamic Deviation (FDD) Metrics for 3D Talking Heads Evaluation
## 🚀 Feature

I would like to propose incorporating an essential evaluation metric for 3D talking heads into the TorchMetrics library:  Upper Face Dynamic Deviation (FDD) . This metric is widely used in speech-driven facial animation research, it measures the variation of facial dynamics for motion sequences in comparison with ground truth. It gives an indication of how close the standard deviation (or upper face motion variation) of generated sequences (of test-set audios) is compared to the variation observed in ground truth.

### Motivation

Current TorchMetrics offerings lack dedicated metrics for evaluating 3D talking heads, except for [LVE](https://github.com/Lightning-AI/torchmetrics/issues/3003). I think this metric also fits in multimodal folder of this library.

### Pitch

Image

### Reference
 
- Paper : [Codetalker](https://openaccess.thecvf.com/content/CVPR2023/papers/Xing_CodeTalker_Speech-Driven_3D_Facial_Animation_With_Discrete_Motion_Prior_CVPR_2023_paper.pdf)

### Additional context

If agreed, I would like to open a PR for the same.

### PR DESCRIPTION
**Pull request by VijayVignesh1**
## What does this PR do?
Implements the Upper Face Dynamics Deviation (FDD) metric for 3D talking head evaluation.
Computes the deviation in motion magnitude of upper-face vertices between predicted and ground truth meshes.

Fixes #3097 

<details>
  <summary>Before submitting</summary>

- [x] Was this **discussed/agreed** via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/torchmetrics/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure to **update the docs**?
- [x] Did you write any new **necessary tests**?

</details>

<details>
  <summary>PR review</summary>

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

</details>

## Did you have fun?

Make sure you had fun coding 🙃

----
📚 Documentation preview 📚: https://torchmetrics--3238.org.readthedocs.build/en/3238/

### CODE REVIEW

File: docs/source/links.rst [hunk 1]
```diff
@@ -1,6 +1,6 @@
 
-.. _scikit-learn's implementation of SMAPE: https://scikit-learn.org/stable/modules/generated/sklearn.metrics.mean_absolute_percentage_error.html
-.. _scikit-learn's implementation of MAPE: https://scikit-learn.org/stable/modules/generated/sklearn.metrics.mean_absolute_percentage_error.html
+.. _scikit-learn's implementation of SMAPE: https://scikit-learn.org/1.7/modules/generated/sklearn.metrics.mean_absolute_percentage_error.html
```
Comment (Borda) [inline docs/source/links.rst]: these seem to to be artefacts after build docs locally

Comment (VijayVignesh1) [reply docs/source/links.rst]: Changed it.

File: src/torchmetrics/functional/multimodal/fdd.py [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+            f"upper_face_map contains invalid vertex indices. Max index {max(upper_face_map)} is larger than "
+            f"number of vertices {vertices_pred.shape[1]}."
+        )
+    if vertices_pred.shape[0] < 2:
+        raise ValueError("Need at least 2 frames to compute dynamics deviation.")
+
+    pred = vertices_pred[:, upper_face_map, :]  # (T, M, 3)
+    gt = vertices_gt[:, upper_face_map, :]
+
+    pred_disp = pred[1:] - pred[:-1]  # (T-1, M, 3)
```
Comment (rittik9) [inline src/torchmetrics/functional/multimodal/fdd.py]: disp should be corresponding to Template mesh of Shape V*3 , which user will be passing. ref: [codetalker](https://github.com/Doubiiu/CodeTalker/blob/main/main/cal_metric.py)

Comment (VijayVignesh1) [reply src/torchmetrics/functional/multimodal/fdd.py]: Done! I have changed the implementation to include the template from the user.

File: src/torchmetrics/functional/multimodal/fdd.py [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+        )
+    if not upper_face_map:
+        raise ValueError("upper_face_map cannot be empty.")
+    if max(upper_face_map) >= vertices_pred.shape[1]:
+        raise ValueError(
+            f"upper_face_map contains invalid vertex indices. Max index {max(upper_face_map)} is larger than "
+            f"number of vertices {vertices_pred.shape[1]}."
+        )
+    if vertices_pred.shape[0] < 2:
+        raise ValueError("Need at least 2 frames to compute dynamics deviation.")
```
Comment (rittik9) [inline src/torchmetrics/functional/multimodal/fdd.py]:
```suggestion
```

File: src/torchmetrics/functional/multimodal/fdd.py [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+
+    """
+    if vertices_pred.ndim != 3 or vertices_gt.ndim != 3:
+        raise ValueError(
+            f"Expected both vertices_pred and vertices_gt to have 3 dimensions but got "
+            f"{vertices_pred.ndim} and {vertices_gt.ndim} dimensions respectively."
+        )
+    if template.ndim != 2 or template.shape[1] != 3:
+        raise ValueError(f"Expected template to have shape (V, 3) but got {template.shape}.")
+    if vertices_pred.shape != vertices_gt.shape:
```
Comment (rittik9) [inline src/torchmetrics/functional/multimodal/fdd.py]:
```suggestion
    if vertices_pred.shape[1:] != vertices_gt.shape[1:]:
```

File: src/torchmetrics/functional/multimodal/fdd.py [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+        )
+    if not upper_face_map:
+        raise ValueError("upper_face_map cannot be empty.")
+    if max(upper_face_map) >= vertices_pred.shape[1]:
+        raise ValueError(
+            f"upper_face_map contains invalid vertex indices. Max index {max(upper_face_map)} is larger than "
+            f"number of vertices {vertices_pred.shape[1]}."
+        )
+
+    pred = vertices_pred[:, upper_face_map, :]  # (T, M, 3)
```
Comment (rittik9) [inline src/torchmetrics/functional/multimodal/fdd.py]:
```suggestion
    min_frames = min(vertices_pred.shape[0], vertices_gt.shape[0])
    pred = vertices_pred[:min_frames, upper_face_map, :]  # (T, M, 3)
```

File: src/torchmetrics/functional/multimodal/fdd.py [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+    gt = vertices_gt[:, upper_face_map, :]
```
Comment (rittik9) [inline src/torchmetrics/functional/multimodal/fdd.py]:
```suggestion
    gt = vertices_gt[:min_frames, upper_face_map, :]
```

File: src/torchmetrics/multimodal/fdd.py [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+        template: Tensor,
+        upper_face_map: List[int],
+        **kwargs: Any,
+    ) -> None:
+        super().__init__(**kwargs)
+        self.upper_face_map = upper_face_map
+        self.template = template
+
+        if not self.upper_face_map:
+            raise ValueError("upper_face_map cannot be empty.")
```
Comment (rittik9) [inline src/torchmetrics/multimodal/fdd.py:108]:
```suggestion
        if not self.template:
            raise ValueError("template cannot be empty.")
        if not self.upper_face_map:
            raise ValueError("upper_face_map cannot be empty.")
```

File: src/torchmetrics/multimodal/fdd.py [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+            vertices_gt: Ground truth vertices tensor of shape (T', V, 3) where T is number of frames,
+                V is number of vertices, and 3 represents XYZ coordinates
+
+        """
+        if vertices_pred.ndim != 3 or vertices_gt.ndim != 3:
+            raise ValueError(
+                f"Expected both vertices_pred and vertices_gt to have 3 dimensions but got "
+                f"{vertices_pred.ndim} and {vertices_gt.ndim} dimensions respectively."
+            )
+        if vertices_pred.shape != vertices_gt.shape:
```
Comment (rittik9) [inline src/torchmetrics/multimodal/fdd.py]:
```suggestion
        if vertices_pred.shape[1:] != vertices_gt.shape[1:]:
```

File: src/torchmetrics/multimodal/fdd.py [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+                f"shapes {vertices_pred.shape} and {vertices_gt.shape}."
+            )
+        if max(self.upper_face_map) >= vertices_pred.shape[1]:
+            raise ValueError(
+                f"upper_face_map contains invalid vertex indices. Max index {max(self.upper_face_map)} is larger than "
+                f"number of vertices {vertices_pred.shape[1]}."
+            )
+
+        self.vertices_pred_list.append(vertices_pred)
+        self.vertices_gt_list.append(vertices_gt)
```
Comment (rittik9) [inline src/torchmetrics/multimodal/fdd.py]:
```suggestion
        min_frames = min(vertices_pred.shape[0], vertices_gt.shape[0])
        self.vertices_pred_list.append(vertices_pred[:min_frames])
        self.vertices_gt_list.append(vertices_gt[:min_frames])
```

File: src/torchmetrics/multimodal/fdd.py [hunk 10]
```diff
@@ -0,0 +1,199 @@
...
+        template: Tensor,
+        upper_face_map: List[int],
+        **kwargs: Any,
+    ) -> None:
+        super().__init__(**kwargs)
+        self.upper_face_map = upper_face_map
+        self.template = template
+
+        if not self.template:
+            raise ValueError("template cannot be empty.")
```
Comment (rittik9) [inline src/torchmetrics/multimodal/fdd.py]:
```suggestion
```

Comment (rittik9) [inline src/torchmetrics/multimodal/fdd.py]:
```suggestion
        if self.template.ndim != 2 or self.template.shape[1] != 3:
            raise ValueError(f"Expected template to have shape (V, 3) but got {template.shape}.")
```

File: src/torchmetrics/functional/multimodal/fdd.py (hunk 5 above)
Comment (VijayVignesh1) [reply src/torchmetrics/functional/multimodal/fdd.py]: @rittik9 Is there a reason why we are going with the min_frames logic? preds and gt need to have the same number of frames, right?

**Comment by VijayVignesh1** [comment]
@Borda @rittik9 The build docs failure seems to be coming from a different broken link.

File: src/torchmetrics/functional/multimodal/fdd.py (hunk 5 above)
Comment (rittik9) [reply src/torchmetrics/functional/multimodal/fdd.py]: @VijayVignesh1 not necessarily. Given an audio input a neural network based model can predict different no of frames from no of ground truth frames.

Comment (VijayVignesh1) [reply src/torchmetrics/functional/multimodal/fdd.py]: Gotcha! Thanks for the clarification.

File: src/torchmetrics/multimodal/fdd.py [hunk 11]
```diff
@@ -0,0 +1,199 @@
...
+    in the motion magnitude of upper-face vertices between the generated and ground truth meshes. It quantifies
+    how well the predicted motion dynamics match the ground truth over time.
+
+    The metric is defined as:
+
+    .. math::
+        \text{FDD} = \frac{1}{|\text{SU}|} \sum_{v \in \text{SU}} \Big( \text{std}(\| x_{1:T,v} -
+        \text{template}_v \|_2^2) - \text{std}(\| \hat{x}_{1:T,v} - \text{template}_v \|_2^2) \Big)
+
+    where :math:`T` is the number of frames, :math:`M = |\text{SU}|` is the number of vertices in the upper-face region,
```
[the remark this answers is not in the document]
Comment (VijayVignesh1) [reply src/torchmetrics/multimodal/fdd.py]: Removed the lower bound.

File: src/torchmetrics/multimodal/fdd.py [hunk 12]
```diff
@@ -0,0 +1,199 @@
...
+        **kwargs: Any,
+    ) -> None:
+        super().__init__(**kwargs)
+        self.upper_face_map = upper_face_map
+        self.template = template
+
+        if self.template.ndim != 2 or self.template.shape[1] != 3:
+            raise ValueError(f"Expected template to have shape (V, 3) but got {template.shape}.")
+        if not self.upper_face_map:
+            raise ValueError("upper_face_map cannot be empty.")
```
[the remark this answers is not in the document]
Comment (VijayVignesh1) [reply src/torchmetrics/multimodal/fdd.py:108]: Added this check.

**Review by rittik9 — APPROVED** [review]

File: src/torchmetrics/functional/multimodal/fdd.py [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+        \text{template}_v \|_2^2) - \text{std}(\| \hat{x}_{1:T,v} - \text{template}_v \|_2^2) \Big)
+
+    where :math:`T` is the number of frames, :math:`S_U` is the set of upper-face vertices with :math:`M = |S_U|`,
+    :math:`x_{t,v}` are the 3D coordinates of vertex :math:`v` at frame :math:`t` in the ground truth sequence,
+    and :math:`\hat{x}_{t,v} \in \mathbb{R}^3` are the corresponding predicted vertices. The neutral template coordinate
+    of vertex :math:`v` is denoted as :math:`\text{template}_v \in \mathbb{R}^3`. The operator :math:`\text{std}(\cdot)`
+    computes the standard deviation of the temporal sequence.
+
+    Args:
+        vertices_pred: Predicted vertices tensor of shape (T, V, 3) where T is number of frames,
```
Comment (justusschock) [inline src/torchmetrics/functional/multimodal/fdd.py]: so T is basically the batch-size here? can we name it that way?
also dies this work in 2d as well? for distance calculation it should not matter if it's R^2 or R^3

Comment (VijayVignesh1) [reply src/torchmetrics/functional/multimodal/fdd.py]: T is the time dimension here. But actually, we need to add the batch dimension to it. To answer the second question, yes this should work with 2d as well. We might need to change the corresponding checks here.
@rittik9 I think we might need to pivot. Your thoughts?

Comment (VijayVignesh1) [reply src/torchmetrics/functional/multimodal/fdd.py]: Just some thoughts on this. I can add the time dimension, so that the input to the class and function is (B, T, V, 3) or (B, T, V, 2). However, the min_frames logic needs to be removed in that case and it's upto the user to align the T dimension of pred and gt. @justusschock @rittik9

Comment (rittik9) [reply src/torchmetrics/functional/multimodal/fdd.py]: Yeah sounds good to me!! I don't think we need (B,T,V,2). Let's just keep it for 3D vertices.

**Review by justusschock — APPROVED** [review]

File: src/torchmetrics/functional/multimodal/fdd.py (hunk 13 above)
Comment (VijayVignesh1) [reply src/torchmetrics/functional/multimodal/fdd.py]: I just realized this comment was pending. I sent in a change to add the batch size dimensions along with the updated test.
